Fixes #15854 - Added tests

witlessbird opened the pull request:
Also broke up provider into smaller components to make them more testable

witlessbird commented:
I also:
 - broke up the provider into smaller components to make it more testable
 - switched to use new plugin initialization
 - rewrote host-record related code to make fewer calls to infoblox
 - Used @georg-rath approach to ranges in free ip code 
 - removed a few configuration options: only infoblox_user, infoblox_pw, record_type, and use_ranges remain

witlessbird commented:
fixed a couple of issues found in testing.

iNecas commented on lib/smart_proxy_dhcp_infoblox/plugin_configuration.rb:
```diff
@@ -0,0 +1,37 @@
+      c.dependency :connection, (lambda do
+                                  ::Infoblox.wapi_version = '2.0'
+                                  ::Infoblox::Connection.new(:username => settings[:infoblox_user] ,:password => settings[:infoblox_pw],
+                                                             :host => settings[:server], :ssl_opts => {:verify => false})
```
I guess we should expose the `:ssl_opts` via settings and discourage using `:verify` false.

iNecas commented:
While testing, I've noticed the gemspec is missing dependency on `infoblox`, could you update that as well?

witlessbird commented:
'infoblox' dependency should go into into :dhcp_infoblox group in here: https://github.com/theforeman/smart_proxy_dhcp_infoblox/blob/master/bundler.d/dhcp_infoblox.rb. Otherwise the gem will always be loaded whether infoblox provider is used or not.

iNecas commented:
While I understand the need for this in smart proxy, where there are multiple providers, and it's expected not every one of them would be used, it seems in this case, there is no benefit, as just the whole gem can be uninstalled. Leaving the decision on you: it was just a bit surprising that a gem doesn't take it's dependencies with it in gemspec.

iNecas commented on lib/smart_proxy_dhcp_infoblox/dhcp_infoblox_plugin.rb:
```diff
@@ -1,25 +1,13 @@
-     :range => false,
-     :infoblox_subnets => false,
-     :restart => false
+    default_settings :record_type => 'host', :range => false
```
Where could we validate the record_type values settings, to prevent typos?

iNecas commented on lib/smart_proxy_dhcp_infoblox/grid_restart.rb:
```diff
@@ -0,0 +1,29 @@
+    def try_restart
+      logger.debug 'Restarting grid.'
+
+      3.times do |tries|
```
Should the 3 be extracted to a constant, perhaps a config?

iNecas commented:
I've tested the gem with latest foreman and proxy and everything worked fine. :+1: from functionality point of view. I left some comments inline.

I've tried running tests against 1.8.7, with adding this to Gemfile (+ removing the test-unit from gemspec):
```
if RUBY_VERSION < '1.9.3'
  gem 'rake', '< 11'
else
  gem 'rake'
  gem 'test-unit'
end

if RUBY_VERSION < '2.0.0'
  gem 'json', '< 2.0.0', :require => false
end
```

And getting 2 tests failing.

witlessbird commented:
* Added validation of record_type setting
* Fixed tests failing on 1.8.7
* Renamed settings to username and password
* extracted maximum number of restart attempts of grid service into a constant

iNecas commented:
@Klaas- @sjoeboo any comments regarding this and https://github.com/theforeman/smart_proxy_dns_infoblox/pull/7? We would like to proceed with packaging (https://github.com/theforeman/foreman-packaging/pull/1290) and installer so that we make it for foreman 1.13 branching, which will happen next week. We are getting to a point where this can block the work.

mbacovsky commented:
FYI I've prepared packaging for smart_proxy_dhcp_infoblox [1] in branch waiting with PR for the gem release and created scratch build [2] of the gem freshly built from this branch. As a dependency it needs `infoblox` packaged - PR is here [3]
[1] https://github.com/mbacovsky/foreman-packaging/commit/3ca8ae5b085e0cc72fee40501a9b9003de4ee5e7
[2] https://koji.katello.org/koji/taskinfo?taskID=520163
[3] https://github.com/theforeman/foreman-packaging/pull/1290

sjoeboo commented:
Looks good to me, reviewing the DNS pluggin PR in a moment!

iNecas commented on README.md:
```diff
@@ -23,13 +23,10 @@ To enable this DHCP provider, edit `/etc/foreman-proxy/settings.d/dhcp.yml` and
-* infoblox_pw: API Password
+* username: API Username
+* passowrd: API Password
 * record_type: host / fixed_address (see different record types chapter)
```
we validate `fixedaddress` while here and in setting example, we use term `fixed_address`

iNecas commented:
One inconsistency in the docs, but ACK otherwise: tested and worked well

witlessbird commented:
Fixed inconsistent use of fixedaddress in README.

iNecas commented:
ACK

mbacovsky commented:
While packaging this plugin I noticed that other proxy plugins e.g. [1] have `settings.d` directory while this one has `config`. It is doesn't prevent packaging in any way but I was curious about the reason for the inconsistency.
[1] https://github.com/theforeman/smart_proxy_discovery

witlessbird commented:
Probably no reason, but it also shouldn't affect anything -- the settings directory contains an example config file.

georg-rath commented:
Awesome!
Just tried it out: it works as expected and is noticeably faster.
